R/fonts.R

```diff
@@ -0,0 +1,508 @@
+                        cache = sass_file_cache(sass_cache_context_dir()),
+                        wght = NULL, ital = NULL, display = c("swap", "auto", "block", "fallback", "optional")) {
+  stopifnot(is.logical(local))
+  if (!is.null(wght)) {
+    stopifnot(is.character(wght) || is.numeric(wght))
+    wght <- sort(wght)
+  }
+  if (!is.null(ital)) {
+    stopifnot(all(ital %in% c(0, 1)))
+    ital <- sort(ital)
+  }
+  display <- match.arg(display)
+
+  axis_rng <-
+    if (is.null(wght) && is.null(ital)) {
+      ""
+    } else if (is.null(ital)) {
+      paste0(":wght@", paste0(wght, collapse = ";"))
+    } else if (is.null(wght)) {
+      paste0(":ital@", paste0(ital, collapse = ";"))
+    } else {
+      paste0(":ital,wght@", paste0(
+        apply(expand.grid(wght, ital)[, 2:1], 1, paste0, collapse = ","),
+        collapse = ";"
+      ))
+    }
+
+  x <- list(
+    family = family,
+    local = isTRUE(local), cache = cache,
+    href = paste0(
+      "https://fonts.googleapis.com/css2?family=",
+      family, axis_rng, "&display=", display
+    )
+  )
+
+  dep_func <- if (x$local) font_dep_google_local else font_dep_link
+  font_object(x, dep_func)
+}
+
+#' @rdname font_face
+#' @export
+#' @param href A URL resource pointing to the font data.
+font_link <- function(family, href) {
+  font_object(list(family = family, href = href), font_dep_link)
+}
+
+#' @rdname font_face
+#' @export
+#' @param src A character vector for the `src` `@font-face` property. Beware
+#'   that is character strings are taken verbatim, so careful quoting and/or URL
+#'   encoding may be required.
+#' @param weight A character (or numeric) vector for the `font-weight`
+#'   `@font-face` property.
+#' @param display A character vector for the `font-display` `@font-face`
+#'   property.
+#' @param style A character vector for the `font-style` `@font-face` property.
+#' @param stretch A character vector for the `font-stretch` `@font-face`
+#'   property.
+#' @param variant A character vector for the `font-variant` `@font-face`
+#'   property.
+#' @param unicode_range A character vector for `unicode-range` `@font-face`
+#'   property.
+font_face <- function(family, src, weight = NULL, style = NULL,
+                      display = c("swap", "auto", "block", "fallback", "optional"),
+                      stretch = NULL, variant = NULL, unicode_range = NULL) {
+
+  x <- dropNulls(list(
+    family = quote_css_font_families(family),
+    src = src,
+    weight = weight,
+    style = style,
+    display = if (!is.null(display)) match.arg(display),
+    stretch = stretch,
+    variant = variant,
+    unicode_range = unicode_range
+  ))
+
+  # Multiple src values are separated by "," (everything else by white space)
+  # TODO: src could accept a list of named lists which might give us the
+  # opportunity to handle quoting and encoding of URLs
+  for (prop in names(x)) {
+    collapse <- if (prop %in% c("src", "unicode_range")) ", " else " "
```

Comment:
Let's keep this code as is, as the speed difference isn't bad.


``` r
for (a in c("src", "unicode_range", "not a match")) {
  print(a)
  print(
    microbenchmark::microbenchmark(
      in_ = { 
        val <- if (a %in% c("src", "unicode_range")) "," else " " 
      }, 
      switch = { 
        val <- switch(a, src = , unicode_range = ",", " ") 
      }, 
      ifelse = { 
        val <- if (a == "src") {","} else if (a == "unicode_range") {","} else {" "}
      }
    )
  )
}
#> [1] "src"
#> Unit: nanoseconds
#>    expr min   lq    mean median     uq   max neval cld
#>     in_ 996 1030 1189.25 1050.5 1090.5 10589   100   b
#>  switch 279  286  326.23  292.5  299.0  3117   100  a 
#>  ifelse 385  408  453.64  430.0  454.0  2356   100  a 
#> [1] "unicode_range"
#> Unit: nanoseconds
#>    expr  min     lq    mean median     uq   max neval cld
#>     in_ 1018 1052.0 1342.95 1068.0 1128.5 12747   100   b
#>  switch  281  314.0  412.01  318.5  324.0  8300   100  a 
#>  ifelse  628  654.5  710.68  667.0  698.0  2777   100  a 
#> [1] "not a match"
#> Unit: nanoseconds
#>    expr  min     lq    mean median     uq   max neval cld
#>     in_ 1024 1055.5 1272.26 1071.0 1116.0 18536   100   c
#>  switch  268  301.0  321.27  304.0  310.0  1170   100 a  
#>  ifelse  646  665.0  723.33  687.5  714.5  3090   100  b
```

<sup>Created on 2021-03-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

But using a `switch` statement might be _marginally faster_ approach as we run into these situations more. (Trying to avoid "death by 1000 cuts")